File: copernicus-official/clients/copernicus-official-client.js
/**
 * Copernicus Official Client - Frontend JavaScript
 * Cliente JavaScript para integração com APIs oficiais do Copernicus
 * Baseado na documentação oficial do Copernicus Data Space Ecosystem
 */

class CopernicusOfficialClient {
    constructor(workerUrl = '/api') {
        this.workerUrl = workerUrl;
        this.accessToken = null;
        this.tokenExpires = null;
        
        // Configuração oficial
        this.config = {
            collections: {
                'SENTINEL-1': 'SENTINEL-1',
                'SENTINEL-2': 'SENTINEL-2',
                'SENTINEL-3': 'SENTINEL-3',
                'SENTINEL-5P': 'SENTINEL-5P',
                'SENTINEL-6': 'SENTINEL-6'
            },
            angolaEEZ: {
                north: -4.2,
                south: -18.0,
                east: 17.5,
                west: 8.5
            }
        };
        
        // Cache de dados
        this.cache = new Map();
        this.cacheTimeout = 5 * 60 * 1000; // 5 minutos
        
        console.log('🛰️ Copernicus Official Client inicializado');
    }
    
    /**
     * Autenticar com Copernicus CDSE
     */
    async authenticate() {
        try {
            console.log('🔐 Autenticando com Copernicus CDSE...');
            
            const response = await fetch(`${this.workerUrl}/copernicus/auth`, {
                method: 'GET',
                headers: {
                    'Accept': 'application/json'
                }
            });
            
            if (!response.ok) {
                throw new Error(`Autenticação falhou: ${response.status}`);
            }
            
            const data = await response.json();
            
            if (data.success) {
                this.accessToken = data.access_token;
                this.tokenExpires = Date.now() + (data.expires_in * 1000);
                
                console.log('✅ Autenticação CDSE bem-sucedida');
                this.showNotification('✅ Copernicus conectado', 'success');
                return true;
            } else {
                throw new Error(data.error || 'Falha na autenticação');
            }
            
        } catch (error) {
            console.error('❌ Erro na autenticação:', error);
            this.showNotification(`❌ Erro de autenticação: ${error.message}`, 'error');
            return false;
        }
    }
    
    /**
     * Verificar se o token ainda é válido
     */
    isTokenValid() {
        return this.accessToken && this.tokenExpires && Date.now() < this.tokenExpires;
    }
    
    /**
     * Buscar dados usando OData API
     */
    async searchOData(collection = 'SENTINEL-3', maxRecords = 20, days = 7) {
        const cacheKey = `odata_${collection}_${maxRecords}_${days}`;
        
        // Verificar cache
        if (this.cache.has(cacheKey)) {
            const cached = this.cache.get(cacheKey);
            if (Date.now() - cached.timestamp < this.cacheTimeout) {
                console.log('📋 Usando dados OData do cache');
                return cached.data;
            }
        }
        
        try {
            console.log(`🔍 Buscando dados OData: ${collection}`);
            
            const params = new URLSearchParams({
                collection,
                max_records: maxRecords.toString(),
                days: days.toString()
            });
            
            const response = await fetch(`${this.workerUrl}/copernicus/odata?${params}`, {
                method: 'GET',
                headers: {
                    'Accept': 'application/json'
                }
            });
            
            if (!response.ok) {
                throw new Error(`OData API falhou: ${response.status}`);
            }
            
            const data = await response.json();
            
            if (data.success) {
                // Cachear resultado
                this.cache.set(cacheKey, {
                    data: data,
                    timestamp: Date.now()
                });
                
                console.log(`✅ OData: ${data.products_found} produtos encontrados`);
                return data;
            } else {
                throw new Error(data.error || 'Erro na busca OData');
            }
            
        } catch (error) {
            console.error('❌ Erro na busca OData:', error);
            this.showNotification(`❌ Erro OData: ${error.message}`, 'error');
            return null;
        }
    }
    
    /**
     * Buscar dados usando STAC API
     */
    async searchSTAC(collections = ['SENTINEL-3'], limit = 50, days = 7) {
        const cacheKey = `stac_${collections.join(',')}_${limit}_${days}`;
        
        // Verificar cache
        if (this.cache.has(cacheKey)) {
            const cached = this.cache.get(cacheKey);
            if (Date.now() - cached.timestamp < this.cacheTimeout) {
                console.log('📋 Usando dados STAC do cache');
                return cached.data;
            }
        }
        
        try {
            console.log(`🔍 Buscando dados STAC: ${collections.join(', ')}`);
            
            const params = new URLSearchParams({
                collections: collections.join(','),
                limit: limit.toString(),
                days: days.toString()
            });
            
            const response = await fetch(`${this.workerUrl}/copernicus/stac?${params}`, {
                method: 'GET',
                headers: {
                    'Accept': 'application/json'
                }
            });
            
            if (!response.ok) {
                throw new Error(`STAC API falhou: ${response.status}`);
            }
            
            const data = await response.json();
            
            if (data.success) {
                // Cachear resultado
                this.cache.set(cacheKey, {
                    data: data,
                    timestamp: Date.now()
                });
                
                console.log(`✅ STAC: ${data.features_found} features encontradas`);
                return data;
            } else {
                throw new Error(data.error || 'Erro na busca STAC');
            }
            
        } catch (error) {
            console.error('❌ Erro na busca STAC:', error);
            this.showNotification(`❌ Erro STAC: ${error.message}`, 'error');
            return null;
        }
    }
    
    /**
     * Buscar dados usando OpenSearch API
     */
    async searchOpenSearch(collection = 'Sentinel3', maxRecords = 20, days = 7) {
        const cacheKey = `opensearch_${collection}_${maxRecords}_${days}`;
        
        // Verificar cache
        if (this.cache.has(cacheKey)) {
            const cached = this.cache.get(cacheKey);
            if (Date.now() - cached.timestamp < this.cacheTimeout) {
                console.log('📋 Usando dados OpenSearch do cache');
                return cached.data;
            }
        }
        
        try {
            console.log(`🔍 Buscando dados OpenSearch: ${collection}`);
            
            const params = new URLSearchParams({
                collection,
                max_records: maxRecords.toString(),
                days: days.toString()
            });
            
            const response = await fetch(`${this.workerUrl}/copernicus/opensearch?${params}`, {
                method: 'GET',
                headers: {
                    'Accept': 'application/json'
                }
            });
            
            if (!response.ok) {
                throw new Error(`OpenSearch API falhou: ${response.status}`);
            }
            
            const data = await response.json();
            
            if (data.success) {
                // Cachear resultado
                this.cache.set(cacheKey, {
                    data: data,
                    timestamp: Date.now()
                });
                
                console.log(`✅ OpenSearch: ${data.products_found} produtos encontrados`);
                return data;
            } else {
                throw new Error(data.error || 'Erro na busca OpenSearch');
            }
            
        } catch (error) {
            console.error('❌ Erro na busca OpenSearch:', error);
            this.showNotification(`❌ Erro OpenSearch: ${error.message}`, 'error');
            return null;
        }
    }
    
    /**
     * Obter dados marinhos completos para Angola
     */
    async getAngolaMarineData() {
        const cacheKey = 'angola_marine_complete';
        
        // Verificar cache
        if (this.cache.has(cacheKey)) {
            const cached = this.cache.get(cacheKey);
            if (Date.now() - cached.timestamp < this.cacheTimeout) {
                console.log('📋 Usando dados marinhos do cache');
                return cached.data;
            }
        }
        
        try {
            console.log('🌊 Buscando dados marinhos completos para Angola...');
            
            const response = await fetch(`${this.workerUrl}/copernicus/angola-marine`, {
                method: 'GET',
                headers: {
                    'Accept': 'application/json'
                }
            });
            
            if (!response.ok) {
                throw new Error(`API Angola falhou: ${response.status}`);
            }
            
            const data = await response.json();
            
            // Cachear resultado
            this.cache.set(cacheKey, {
                data: data,
                timestamp: Date.now()
            });
            
            console.log(`✅ Dados Angola: ${data.summary?.total_products_found || 0} produtos total`);
            this.showNotification(`✅ Dados Angola atualizados: ${data.summary?.total_products_found || 0} produtos`, 'success');
            
            return data;
            
        } catch (error) {
            console.error('❌ Erro nos dados marinhos Angola:', error);
            this.showNotification(`❌ Erro dados Angola: ${error.message}`, 'error');
            return null;
        }
    }
    
    /**
     * Processar dados Sentinel-3 para visualização
     */
    processMarineDataForVisualization(data) {
        if (!data || !data.data_sources) {
            return null;
        }
        
        const processed = {
            timestamp: data.timestamp,
            angola_eez: data.angola_eez,
            summary: data.summary,
            oceanographic_data: {
                temperature: [],
                chlorophyll: [],
                salinity: [],
                currents: []
            },
            metadata: {
                total_products: 0,
                successful_apis: 0,
                data_sources: []
            }
        };
        
        // Processar dados de cada fonte
        Object.entries(data.data_sources).forEach(([source, sourceData]) => {
            if (!sourceData.error) {
                processed.metadata.successful_apis++;
                processed.metadata.data_sources.push(source);
                
                // Processar produtos/features
                const items = sourceData.data?.value || sourceData.data?.features || [];
                processed.metadata.total_products += items.length;
                
                items.forEach(item => {
                    // Extrair dados oceanográficos baseado no nome do produto
                    const name = item.Name || item.id || '';
                    
                    if (name.includes('OLCI')) {
                        // Dados de cor do oceano (clorofila)
                        processed.oceanographic_data.chlorophyll.push({
                            id: item.Id || item.id,
                            name: name,
                            date: item.ContentDate?.Start || item.properties?.datetime,
                            geometry: item.GeoFootprint || item.geometry
                        });
                    } else if (name.includes('SLSTR')) {
                        // Dados de temperatura da superfície
                        processed.oceanographic_data.temperature.push({
                            id: item.Id || item.id,
                            name: name,
                            date: item.ContentDate?.Start || item.properties?.datetime,
                            geometry: item.GeoFootprint || item.geometry
                        });
                    }
                });
            }
        });
        
        return processed;
    }
    
    /**
     * Atualizar status na interface
     */
    updateCopernicusStatus(status, message) {
        const statusElement = document.getElementById('copernicus-status');
        const statusText = document.getElementById('copernicus-status-text');
        
        if (statusElement && statusText) {
            statusElement.className = `status-indicator ${status}`;
            statusText.textContent = message;
            
            // Cores baseadas no status
            switch (status) {
                case 'online':
                    statusElement.style.backgroundColor = '#10b981'; // Verde
                    break;
                case 'warning':
                    statusElement.style.backgroundColor = '#f59e0b'; // Amarelo
                    break;
                case 'error':
                    statusElement.style.backgroundColor = '#ef4444'; // Vermelho
                    break;
                default:
                    statusElement.style.backgroundColor = '#6b7280'; // Cinza
            }
        }
        
        console.log(`📊 Status Copernicus: ${status} - ${message}`);
    }
    
    /**
     * Mostrar notificação
     */
    showNotification(message, type = 'info') {
        // Criar elemento de notificação se não existir
        let notificationContainer = document.getElementById('copernicus-notifications');
        if (!notificationContainer) {
            notificationContainer = document.createElement('div');
            notificationContainer.id = 'copernicus-notifications';
            notificationContainer.style.cssText = `
                position: fixed;
                top: 20px;
                right: 20px;
                z-index: 10000;
                max-width: 400px;
            `;
            document.body.appendChild(notificationContainer);
        }
        
        // Criar notificação
        const notification = document.createElement('div');
        notification.style.cssText = `
            background: ${type === 'success' ? '#10b981' : type === 'error' ? '#ef4444' : '#3b82f6'};
            color: white;
            padding: 12px 16px;
            border-radius: 8px;
            margin-bottom: 8px;
            box-shadow: 0 4px 12px rgba(0,0,0,0.15);
            animation: slideIn 0.3s ease-out;
        `;
        notification.textContent = message;
        
        // Adicionar animação CSS
        if (!document.getElementById('copernicus-notification-styles')) {
            const styles = document.createElement('style');
            styles.id = 'copernicus-notification-styles';
            styles.textContent = `
                @keyframes slideIn {
                    from { transform: translateX(100%); opacity: 0; }
                    to { transform: translateX(0); opacity: 1; }
                }
            `;
            document.head.appendChild(styles);
        }
        
        notificationContainer.appendChild(notification);
        
        // Remover após 5 segundos
        setTimeout(() => {
            if (notification.parentNode) {
                notification.parentNode.removeChild(notification);
            }
        }, 5000);
    }
    
    /**
     * Limpar cache
     */
    clearCache() {
        this.cache.clear();
        console.log('🗑️ Cache Copernicus limpo');
    }
    
    /**
     * Obter estatísticas do cache
     */
    getCacheStats() {
        return {
            size: this.cache.size,
            keys: Array.from(this.cache.keys()),
            totalMemory: JSON.stringify(Array.from(this.cache.values())).length
        };
    }
    
    /**
     * Inicialização automática
     */
    async initialize() {
        console.log('🚀 Inicializando Copernicus Official Client...');
        
        try {
            // Tentar autenticação
            const authSuccess = await this.authenticate();
            
            if (authSuccess) {
                this.updateCopernicusStatus('online', 'Copernicus CDSE Conectado');
                
                // Buscar dados iniciais
                const angolaData = await this.getAngolaMarineData();
                
                if (angolaData) {
                    console.log('✅ Inicialização completa com sucesso');
                    return angolaData;
                }
            }
            
            throw new Error('Falha na inicialização');
            
        } catch (error) {
            console.error('❌ Erro na inicialização:', error);
            this.updateCopernicusStatus('error', 'Erro na Conexão');
            return null;
        }
    }
}

// Instância global
window.CopernicusOfficialClient = CopernicusOfficialClient;

// Auto-inicialização quando DOM estiver pronto
if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', () => {
        window.copernicusClient = new CopernicusOfficialClient();
    });
} else {
    window.copernicusClient = new CopernicusOfficialClient();
}

console.log('🛰️ Copernicus Official Client carregado');

File: copernicus-official/clients/copernicus_official_client.py
#!/usr/bin/env python3
"""
Copernicus Official Client - Implementação correta baseada na documentação oficial
Integração completa com Copernicus Data Space Ecosystem usando APIs oficiais
"""

import requests
import json
import os
from datetime import datetime, timedelta
from typing import Dict, List, Any, Optional
from dataclasses import dataclass
import logging

logger = logging.getLogger(__name__)

@dataclass
class CopernicusConfig:
    """Configuração oficial do Copernicus Data Space Ecosystem"""
    
    # URLs oficiais baseadas na documentação
    IDENTITY_URL = "https://identity.dataspace.copernicus.eu/auth/realms/CDSE/protocol/openid-connect/token"
    CATALOG_ODATA_URL = "https://catalogue.dataspace.copernicus.eu/odata/v1/Products"
    STAC_URL = "https://catalogue.dataspace.copernicus.eu/stac"
    OPENSEARCH_URL = "https://catalogue.dataspace.copernicus.eu/resto/api/collections/search.json"
    DOWNLOAD_URL = "https://catalogue.dataspace.copernicus.eu/odata/v1/Products"
    
    # Configuração de autenticação oficial
    CLIENT_ID = "cdse-public"
    
    # Coleções disponíveis (baseado na documentação)
    COLLECTIONS = {
        'SENTINEL-1': 'SENTINEL-1',
        'SENTINEL-2': 'SENTINEL-2', 
        'SENTINEL-3': 'SENTINEL-3',
        'SENTINEL-5P': 'SENTINEL-5P',
        'SENTINEL-6': 'SENTINEL-6'
    }
    
    # ZEE Angola (coordenadas corretas)
    ANGOLA_EEZ = {
        'north': -4.2,
        'south': -18.0,
        'east': 17.5,
        'west': 8.5
    }


class CopernicusOfficialClient:
    """
    Cliente oficial para Copernicus Data Space Ecosystem
    Implementação baseada na documentação oficial
    """
    
    def __init__(self, username: str = None, password: str = None):
        self.config = CopernicusConfig()
        
        # Credenciais do ambiente ou parâmetros
        self.username = username or os.getenv('COPERNICUS_USERNAME')
        self.password = password or os.getenv('COPERNICUS_PASSWORD')
        
        # Sessão HTTP
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'BGAPP-Angola/2.0 Copernicus-CDSE-Client',
            'Accept': 'application/json'
        })
        
        # Token de acesso
        self.access_token = None
        self.token_expires = None
    
    def authenticate(self) -> bool:
        """
        Autenticação oficial com Copernicus Data Space Ecosystem
        Baseado na documentação: https://documentation.dataspace.copernicus.eu/APIs.html
        """
        try:
            auth_data = {
                'client_id': self.config.CLIENT_ID,
                'grant_type': 'password',
                'username': self.username,
                'password': self.password
            }
            
            logger.info(f"🔐 Autenticando com Copernicus CDSE: {self.username}")
            
            response = self.session.post(
                self.config.IDENTITY_URL,
                data=auth_data,
                headers={'Content-Type': 'application/x-www-form-urlencoded'},
                timeout=30
            )
            
            if response.status_code == 200:
                token_data = response.json()
                self.access_token = token_data['access_token']
                expires_in = token_data.get('expires_in', 3600)
                self.token_expires = datetime.now() + timedelta(seconds=expires_in)
                
                # Configurar header de autorização
                self.session.headers['Authorization'] = f'Bearer {self.access_token}'
                
                logger.info("✅ Autenticação CDSE bem-sucedida!")
                return True
            else:
                logger.error(f"❌ Falha na autenticação CDSE: {response.status_code}")
                logger.error(f"Resposta: {response.text}")
                return False
                
        except Exception as e:
            logger.error(f"❌ Erro na autenticação CDSE: {e}")
            return False
    
    def search_odata(self, collection: str = 'SENTINEL-3', 
                     start_date: datetime = None, 
                     end_date: datetime = None,
                     max_records: int = 20) -> Dict[str, Any]:
        """
        Busca usando OData API (oficial)
        Baseado na documentação: https://documentation.dataspace.copernicus.eu/APIs/OData.html
        """
        if not self.access_token:
            raise Exception("Não autenticado. Execute authenticate() primeiro.")
        
        # Datas padrão (última semana)
        if not end_date:
            end_date = datetime.now()
        if not start_date:
            start_date = end_date - timedelta(days=7)
        
        # Construir filtro OData conforme documentação
        filters = [
            f"Collection/Name eq '{collection}'",
            f"ContentDate/Start gt {start_date.strftime('%Y-%m-%dT%H:%M:%S.000Z')}",
            f"ContentDate/Start lt {end_date.strftime('%Y-%m-%dT%H:%M:%S.000Z')}"
        ]
        
        # Filtro geográfico para Angola (usando sintaxe correta da documentação)
        angola_polygon = f"SRID=4326;POLYGON(({self.config.ANGOLA_EEZ['west']} {self.config.ANGOLA_EEZ['south']}, {self.config.ANGOLA_EEZ['east']} {self.config.ANGOLA_EEZ['south']}, {self.config.ANGOLA_EEZ['east']} {self.config.ANGOLA_EEZ['north']}, {self.config.ANGOLA_EEZ['west']} {self.config.ANGOLA_EEZ['north']}, {self.config.ANGOLA_EEZ['west']} {self.config.ANGOLA_EEZ['south']}))"
        filters.append(f"OData.CSC.Intersects(area=geography'{angola_polygon}')")
        
        # Parâmetros OData
        params = {
            '$filter': ' and '.join(filters),
            '$orderby': 'ContentDate/Start desc',
            '$top': str(max_records),
            '$select': 'Id,Name,ContentDate,GeoFootprint,S3Path,ContentLength'
        }
        
        try:
            logger.info(f"🔍 Buscando dados OData para {collection}")
            
            response = self.session.get(
                self.config.CATALOG_ODATA_URL,
                params=params,
                timeout=60
            )
            
            if response.status_code == 200:
                data = response.json()
                logger.info(f"✅ Encontrados {len(data.get('value', []))} produtos")
                return data
            else:
                logger.error(f"❌ Erro na busca OData: {response.status_code}")
                logger.error(f"Resposta: {response.text}")
                return {'value': [], 'error': response.text}
                
        except Exception as e:
            logger.error(f"❌ Erro na busca OData: {e}")
            return {'value': [], 'error': str(e)}
    
    def search_stac(self, collections: List[str] = ['SENTINEL-3'],
                    start_date: datetime = None,
                    end_date: datetime = None,
                    limit: int = 50) -> Dict[str, Any]:
        """
        Busca usando STAC API (oficial)
        Baseado na documentação: https://documentation.dataspace.copernicus.eu/APIs/STAC.html
        """
        if not self.access_token:
            raise Exception("Não autenticado. Execute authenticate() primeiro.")
        
        # Datas padrão
        if not end_date:
            end_date = datetime.now()
        if not start_date:
            start_date = end_date - timedelta(days=7)
        
        # Corpo da requisição STAC
        search_body = {
            "collections": collections,
            "datetime": f"{start_date.isoformat()}Z/{end_date.isoformat()}Z",
            "bbox": [
                self.config.ANGOLA_EEZ['west'],
                self.config.ANGOLA_EEZ['south'],
                self.config.ANGOLA_EEZ['east'],
                self.config.ANGOLA_EEZ['north']
            ],
            "limit": limit,
            "fields": {
                "include": ["id", "type", "geometry", "bbox", "properties", "assets"],
                "exclude": []
            }
        }
        
        try:
            logger.info(f"🔍 Buscando dados STAC para {collections}")
            
            response = self.session.post(
                f"{self.config.STAC_URL}/search",
                json=search_body,
                headers={'Content-Type': 'application/json'},
                timeout=60
            )
            
            if response.status_code == 200:
                data = response.json()
                logger.info(f"✅ STAC: Encontrados {len(data.get('features', []))} itens")
                return data
            else:
                logger.error(f"❌ Erro na busca STAC: {response.status_code}")
                logger.error(f"Resposta: {response.text}")
                return {'features': [], 'error': response.text}
                
        except Exception as e:
            logger.error(f"❌ Erro na busca STAC: {e}")
            return {'features': [], 'error': str(e)}
    
    def search_opensearch(self, collection: str = 'Sentinel3',
                         start_date: datetime = None,
                         end_date: datetime = None,
                         max_records: int = 20) -> Dict[str, Any]:
        """
        Busca usando OpenSearch API (oficial)  
        Baseado na documentação: https://documentation.dataspace.copernicus.eu/APIs/OpenSearch.html
        """
        # Datas padrão
        if not end_date:
            end_date = datetime.now()
        if not start_date:
            start_date = end_date - timedelta(days=7)
        
        # Parâmetros OpenSearch
        params = {
            'startDate': start_date.strftime('%Y-%m-%dT%H:%M:%S.000Z'),
            'completionDate': end_date.strftime('%Y-%m-%dT%H:%M:%S.000Z'),
            'box': f"{self.config.ANGOLA_EEZ['west']},{self.config.ANGOLA_EEZ['south']},{self.config.ANGOLA_EEZ['east']},{self.config.ANGOLA_EEZ['north']}",
            'maxRecords': str(max_records),
            'sortParam': 'startDate',
            'sortOrder': 'descending'
        }
        
        try:
            logger.info(f"🔍 Buscando dados OpenSearch para {collection}")
            
            # OpenSearch não requer autenticação para busca
            response = requests.get(
                f"{self.config.OPENSEARCH_URL.replace('/search.json', f'/{collection}/search.json')}",
                params=params,
                timeout=60
            )
            
            if response.status_code == 200:
                data = response.json()
                logger.info(f"✅ OpenSearch: Encontrados {len(data.get('features', []))} produtos")
                return data
            else:
                logger.error(f"❌ Erro na busca OpenSearch: {response.status_code}")
                return {'features': [], 'error': response.text}
                
        except Exception as e:
            logger.error(f"❌ Erro na busca OpenSearch: {e}")
            return {'features': [], 'error': str(e)}
    
    def download_product(self, product_id: str, output_path: str) -> bool:
        """
        Download de produto usando ID
        Baseado na documentação: https://documentation.dataspace.copernicus.eu/APIs/OData.html#product-download
        """
        if not self.access_token:
            raise Exception("Não autenticado. Execute authenticate() primeiro.")
        
        download_url = f"{self.config.DOWNLOAD_URL}({product_id})/$value"
        
        try:
            logger.info(f"📥 Baixando produto: {product_id}")
            
            response = self.session.get(
                download_url,
                stream=True,
                timeout=300
            )
            
            if response.status_code == 200:
                with open(output_path, 'wb') as f:
                    for chunk in response.iter_content(chunk_size=8192):
                        if chunk:
                            f.write(chunk)
                
                logger.info(f"✅ Download concluído: {output_path}")
                return True
            else:
                logger.error(f"❌ Erro no download: {response.status_code}")
                return False
                
        except Exception as e:
            logger.error(f"❌ Erro no download: {e}")
            return False
    
    def get_angola_marine_data(self) -> Dict[str, Any]:
        """
        Obter dados marinhos específicos para Angola
        Combina múltiplas APIs para dados completos
        """
        if not self.authenticate():
            return {'error': 'Falha na autenticação'}
        
        results = {
            'timestamp': datetime.now().isoformat(),
            'angola_eez': self.config.ANGOLA_EEZ,
            'data_sources': {}
        }
        
        # 1. Buscar dados Sentinel-3 (oceanográficos)
        s3_data = self.search_odata('SENTINEL-3', max_records=10)
        results['data_sources']['sentinel3_odata'] = {
            'products_found': len(s3_data.get('value', [])),
            'data': s3_data
        }
        
        # 2. Buscar via STAC para metadados ricos
        stac_data = self.search_stac(['SENTINEL-3'], limit=10)
        results['data_sources']['sentinel3_stac'] = {
            'features_found': len(stac_data.get('features', [])),
            'data': stac_data
        }
        
        # 3. Buscar via OpenSearch para compatibilidade
        opensearch_data = self.search_opensearch('Sentinel3', max_records=10)
        results['data_sources']['sentinel3_opensearch'] = {
            'products_found': len(opensearch_data.get('features', [])),
            'data': opensearch_data
        }
        
        return results


# Instância global do cliente oficial
copernicus_client = CopernicusOfficialClient()
